File: packages/hexfield/python/hexfield/features.py
"""Node features (F = 15) from stored position facts.

Expands feature rows from raw facts (the hexfield_compact_v1 columns): the
placement history (q, r, owner, placement_index), phase, first stone, and the
side-relative hot / standing-win cell lists. Indices 0-10 and 12 follow the
dense_cnn plane semantics; index 11 is distance-to-nearest-stone; 13-14 are
the standing-win planes.

Turn structure is deterministic (1-then-2-2-2...), so each history record's
phase and player are derived from its ordinal position; the shard schema
stores neither.
"""

from __future__ import annotations

from dataclasses import dataclass, replace

import numpy as np

from .constants import (
    F_DIST_TO_STONE,
    F_EMPTY,
    F_FIRST_STONE,
    F_LEGAL,
    F_OPP_HOT,
    F_OPP_LAST_TURN,
    F_OPP_RECENCY,
    F_OPP_STONE,
    F_OPP_WIN_NOW,
    F_OWN_HOT,
    F_OWN_RECENCY,
    F_OWN_STONE,
    F_OWN_WIN_NOW,
    F_PHASE_SECOND,
    F_PLAYER_COLOUR,
    DIST_SCALE,
    HOT_MIN_COUNT,
    HOT_MIN_PLACEMENTS,
    NUM_FEATURES,
    WIN_NOW_COUNT,
    WINDOW_LEN,
)
from .geometry import apply_d6
from .support import Support, build_support

PHASE_OPENING = "Opening"
PHASE_FIRST = "FirstStone"
PHASE_SECOND = "SecondStone"

# Win axes: Q=(1,0), R=(0,1), QR=(1,-1).
AXIS_DELTAS: dict[str, tuple[int, int]] = {"Q": (1, 0), "R": (0, 1), "QR": (1, -1)}


@dataclass(frozen=True)
class PositionFacts:
    """Raw facts for one decision state, side-relative where applicable.

    records: chronological placement history (q, r, owner, placement_index);
    owner is 0/1 (player0/player1). placements_made == len(records). Hot / win
    lists are relative to ``current_player``.
    """

    records: tuple[tuple[int, int, int, int], ...]
    current_player: int
    phase: str
    first_stone: tuple[int, int] | None
    own_hot: tuple[tuple[int, int], ...]
    opp_hot: tuple[tuple[int, int], ...]
    own_win: tuple[tuple[int, int], ...]
    opp_win: tuple[tuple[int, int], ...]

    @property
    def placements_made(self) -> int:
        return len(self.records)

    def stones(self) -> list[tuple[int, int]]:
        return [(q, r) for q, r, _owner, _idx in self.records]


def record_phase(ordinal: int) -> str:
    """Phase of the history record at chronological position ``ordinal``."""

    if ordinal == 0:
        return PHASE_OPENING
    return PHASE_FIRST if (ordinal - 1) % 2 == 0 else PHASE_SECOND


def record_player(ordinal: int) -> int:
    """Player (0/1) of the history record at position ``ordinal``."""

    if ordinal == 0:
        return 0
    return 1 if ((ordinal - 1) // 2) % 2 == 0 else 0


# --- window scan (threat/standing-win logic) ------------------------------------


def window_scan(
    records: tuple[tuple[int, int, int, int], ...],
    current_player: int,
    placements_made: int,
) -> tuple[
    tuple[tuple[int, int], ...],
    tuple[tuple[int, int], ...],
    tuple[tuple[int, int], ...],
    tuple[tuple[int, int], ...],
]:
    """(own_hot, opp_hot, own_win, opp_win) cell lists from raw stones.

    Scans single-colour windows only. A window is hot when its stone count is
    >= HOT_MIN_COUNT and placements_made >= HOT_MIN_PLACEMENTS; its empty cells
    are marked hot. A window is a standing win when its stone count ==
    WIN_NOW_COUNT (exactly one empty cell); its empty cell is marked win.
    Hot/win cells are assigned to own_* or opp_* by comparing the window
    owner against ``current_player``.
    """

    owner_at: dict[tuple[int, int], int] = {(q, r): owner for q, r, owner, _ in records}
    own_hot: set[tuple[int, int]] = set()
    opp_hot: set[tuple[int, int]] = set()
    own_win: set[tuple[int, int]] = set()
    opp_win: set[tuple[int, int]] = set()

    seen: set[tuple[int, int, str]] = set()
    for (sq, sr) in owner_at:
        for axis, (dq, dr) in AXIS_DELTAS.items():
            for back in range(WINDOW_LEN):
                start = (sq - back * dq, sr - back * dr)
                key = (start[0], start[1], axis)
                if key in seen:
                    continue
                seen.add(key)
                cells = [
                    (start[0] + i * dq, start[1] + i * dr) for i in range(WINDOW_LEN)
                ]
                counts = [0, 0]
                empties: list[tuple[int, int]] = []
                for cell in cells:
                    owner = owner_at.get(cell)
                    if owner is None:
                        empties.append(cell)
                    else:
                        counts[owner] += 1
                if counts[0] > 0 and counts[1] > 0:
                    continue  # two-coloured window: skipped
                count = counts[0] + counts[1]
                owner = 0 if counts[0] > 0 else 1
                if count == WIN_NOW_COUNT:
                    (own_win if owner == current_player else opp_win).update(empties)
                if count >= HOT_MIN_COUNT and placements_made >= HOT_MIN_PLACEMENTS:
                    (own_hot if owner == current_player else opp_hot).update(empties)

    return (
        tuple(sorted(own_hot)),
        tuple(sorted(opp_hot)),
        tuple(sorted(own_win)),
        tuple(sorted(opp_win)),
    )


# --- feature build ----------------------------------------------------------------


def build_features(facts: PositionFacts, sup: Support) -> np.ndarray:
    """(N, 15) float32 feature matrix in support node order."""

    n = sup.num_nodes
    feats = np.zeros((n, NUM_FEATURES), dtype=np.float32)

    placements_made = facts.placements_made
    for q, r, owner, placement_index in facts.records:
        row = sup.index[(q, r)]
        if owner == facts.current_player:
            feats[row, F_OWN_STONE] = 1.0
            recency_plane = F_OWN_RECENCY
        else:
            feats[row, F_OPP_STONE] = 1.0
            recency_plane = F_OPP_RECENCY
        # age = placements_made - placement_index; recency weight = 1/(1+age).
        age = placements_made - placement_index
        weight = 1.0 / (1.0 + float(age))
        feats[row, recency_plane] = max(feats[row, recency_plane], weight)

    feats[:, F_EMPTY] = 1.0 - feats[:, F_OWN_STONE] - feats[:, F_OPP_STONE]
    feats[: sup.legal_count, F_LEGAL] = 1.0

    if facts.phase == PHASE_SECOND:
        feats[:, F_PHASE_SECOND] = 1.0
        if facts.first_stone is not None:
            feats[sup.index[facts.first_stone], F_FIRST_STONE] = 1.0

    if facts.current_player == 0:
        feats[:, F_PLAYER_COLOUR] = 1.0

    for cell in facts.opp_hot:
        feats[sup.index[cell], F_OPP_HOT] = 1.0
    for cell in facts.own_hot:
        feats[sup.index[cell], F_OWN_HOT] = 1.0
    for cell in facts.opp_win:
        feats[sup.index[cell], F_OPP_WIN_NOW] = 1.0
    for cell in facts.own_win:
        feats[sup.index[cell], F_OWN_WIN_NOW] = 1.0

    feats[:, F_DIST_TO_STONE] = sup.dist.astype(np.float32) / DIST_SCALE

    for cell in _opp_last_turn_cells(facts):
        feats[sup.index[cell], F_OPP_LAST_TURN] = 1.0

    return feats


def _opp_last_turn_cells(facts: PositionFacts) -> list[tuple[int, int]]:
    """Cells of the opponent's most recent full turn.

    Scans history in reverse. The first opponent SecondStone record returns
    both its own cell and the previous record's cell (its first-stone
    companion); an opponent Opening record returns its single cell; opponent
    FirstStone records are skipped."""

    opponent = 1 - facts.current_player
    records = facts.records
    for ordinal in range(len(records) - 1, -1, -1):
        if record_player(ordinal) != opponent:
            continue
        phase = record_phase(ordinal)
        q, r, _owner, _idx = records[ordinal]
        if phase == PHASE_SECOND:
            fq, fr, _o, _i = records[ordinal - 1]
            return [(fq, fr), (q, r)]
        if phase == PHASE_OPENING:
            return [(q, r)]
    return []


def build_position(facts: PositionFacts) -> tuple[Support, np.ndarray]:
    """Support + features for one decision state."""

    sup = build_support(facts.stones())
    return sup, build_features(facts, sup)


# --- D6 augmentation ----------------------------------------------------------------


def transform_facts(facts: PositionFacts, sym: int) -> PositionFacts:
    """Apply D6 transform ``sym`` to every stored coordinate fact.

    Transforms the coordinates in records, first_stone, and the hot/win cell
    lists; all other fields are copied unchanged. Support, node order,
    neighbour table, and features are rebuilt from the transformed facts."""

    def t(cell: tuple[int, int]) -> tuple[int, int]:
        return apply_d6(sym, cell[0], cell[1])

    return replace(
        facts,
        records=tuple(
            (*t((q, r)), owner, idx) for q, r, owner, idx in facts.records
        ),
        first_stone=t(facts.first_stone) if facts.first_stone is not None else None,
        own_hot=tuple(sorted(t(c) for c in facts.own_hot)),
        opp_hot=tuple(sorted(t(c) for c in facts.opp_hot)),
        own_win=tuple(sorted(t(c) for c in facts.own_win)),
        opp_win=tuple(sorted(t(c) for c in facts.opp_win)),
    )

File: packages/shrimp/rust/src/features.rs
//! Node features (F = NUM_FEATURES = 15) computed from the engine state.
//! Index 11 is distance-to-nearest-stone; indices 13-14 are the standing-win
//! planes. A matching implementation lives in python/shrimp/features.py.

use hexo_engine::{Axis, HexCoord, HexoState as RustHexoState, Player, TurnPhase};

use crate::constants::*;
use crate::support::Support;

pub fn axis_delta(axis: Axis) -> (i16, i16) {
    match axis {
        Axis::Q => (1, 0),
        Axis::R => (0, 1),
        Axis::QR => (1, -1),
    }
}

/// (N * NUM_FEATURES) node-major feature matrix in support node order.
pub fn build_features(state: &RustHexoState, sup: &Support) -> Vec<f32> {
    let n = sup.num_nodes();
    let mut feats = vec![0f32; n * NUM_FEATURES];
    let set = |feats: &mut Vec<f32>, row: usize, plane: usize, value: f32| {
        feats[row * NUM_FEATURES + plane] = value;
    };
    let current = state.current_player();
    let placements_made = state.placements_made();

    // Stones and recency. age = placements_made - placement_index (saturating);
    // recency weight = 1/(1+age), max-accumulated per cell.
    for record in state.placement_history().iter() {
        let row = sup.row(record.coord).expect("stone missing from support");
        let (stone_plane, recency_plane) = if record.player == current {
            (F_OWN_STONE, F_OWN_RECENCY)
        } else {
            (F_OPP_STONE, F_OPP_RECENCY)
        };
        set(&mut feats, row, stone_plane, 1.0);
        let age = placements_made.saturating_sub(record.placement_index);
        let weight = 1.0 / (1.0 + age as f32);
        let offset = row * NUM_FEATURES + recency_plane;
        feats[offset] = feats[offset].max(weight);
    }

    for row in 0..n {
        let own = feats[row * NUM_FEATURES + F_OWN_STONE];
        let opp = feats[row * NUM_FEATURES + F_OPP_STONE];
        feats[row * NUM_FEATURES + F_EMPTY] = 1.0 - own - opp;
        feats[row * NUM_FEATURES + F_DIST_TO_STONE] = sup.dist[row] as f32 / DIST_SCALE;
    }
    for row in 0..sup.legal_count {
        set(&mut feats, row, F_LEGAL, 1.0);
    }

    match state.phase() {
        TurnPhase::SecondStone { first } => {
            for row in 0..n {
                set(&mut feats, row, F_PHASE_SECOND, 1.0);
            }
            let row = sup.row(first).expect("first stone missing from support");
            set(&mut feats, row, F_FIRST_STONE, 1.0);
        }
        TurnPhase::Opening | TurnPhase::FirstStone => {}
    }

    if current == Player::Player0 {
        for row in 0..n {
            set(&mut feats, row, F_PLAYER_COLOUR, 1.0);
        }
    }

    fill_hot_and_win(state, current, sup, &mut feats);
    fill_opponent_last_turn(state, current, sup, &mut feats);

    feats
}

/// Hot (count >= 4, gated placements >= 7) and standing-win (count == 5,
/// ungated) planes over single-colour windows' EMPTY cells.
fn fill_hot_and_win(
    state: &RustHexoState,
    current: Player,
    sup: &Support,
    feats: &mut Vec<f32>,
) {
    let hot_enabled = state.placements_made() >= HOT_MIN_PLACEMENTS;
    for entry in state.board().windows().entries() {
        let p0 = entry.mask(Player::Player0);
        let p1 = entry.mask(Player::Player1);
        let c0 = (p0 as u32).count_ones();
        let c1 = (p1 as u32).count_ones();
        if (c0 > 0 && c1 > 0) || (c0 + c1) < HOT_MIN_COUNT {
            continue;
        }
        let count = c0 + c1;
        let owner = if c0 > 0 { Player::Player0 } else { Player::Player1 };
        let key = entry.key();
        let (dq, dr) = axis_delta(key.axis);
        let union = p0 | p1;
        for i in 0..WINDOW_LEN {
            if (union >> i) & 1 != 0 {
                continue;
            }
            let coord = HexCoord {
                q: key.start.q + dq * i as i16,
                r: key.start.r + dr * i as i16,
            };
            let row = sup
                .row(coord)
                .expect("window empty cell missing from support");
            if count == WIN_NOW_COUNT {
                let plane = if owner == current { F_OWN_WIN_NOW } else { F_OPP_WIN_NOW };
                feats[row * NUM_FEATURES + plane] = 1.0;
            }
            if hot_enabled {
                let plane = if owner == current { F_OWN_HOT } else { F_OPP_HOT };
                feats[row * NUM_FEATURES + plane] = 1.0;
            }
        }
    }
}

/// Cells of the opponent's most recent full turn.
fn fill_opponent_last_turn(
    state: &RustHexoState,
    current: Player,
    sup: &Support,
    feats: &mut Vec<f32>,
) {
    let opponent = current.other();
    for record in state.placement_history().iter().rev() {
        if record.player != opponent {
            continue;
        }
        match record.phase {
            TurnPhase::SecondStone { first } => {
                for coord in [first, record.coord] {
                    let row = sup.row(coord).expect("last-turn cell missing from support");
                    feats[row * NUM_FEATURES + F_OPP_LAST_TURN] = 1.0;
                }
                return;
            }
            TurnPhase::Opening => {
                let row = sup
                    .row(record.coord)
                    .expect("last-turn cell missing from support");
                feats[row * NUM_FEATURES + F_OPP_LAST_TURN] = 1.0;
                return;
            }
            TurnPhase::FirstStone => {}
        }
    }
}
